cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,420 @@
+			expectedUpdateResponse: newFakeNotFoundResponse(),
+			expectedUpdateError:    errors.New("failed to create firewall after found missing on update"),
+			expectedError:          errors.New("failed to create firewall after found missing on update"),
+		},
+		{
+			name:                   "return error when failing to add inbound rules on update request to firewall API",
+			fwCache:                newFakeFirewallCache(fakeWorkerFWName, diffFakeInboundRule),
+			inboundRules:           []godo.InboundRule{fakeInboundRule},
+			expectedListResponse:   newFakeOKResponse(),
+			expectedListResult:     newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedUpdateResponse: newFakeOKResponse(),
+			expectedUpdateError:    errors.New("failed to add new firewall inbound rules"),
+			expectedError:          errors.New("failed to add new firewall inbound rules"),
+		},
+		{
+			name:                 "when the firewall cache is nil, return existing firewall from API and update cache",
+			fwCache:              newFakeFirewallCacheEmpty(),
+			inboundRules:         []godo.InboundRule{fakeInboundRule},
+			expectedListResponse: newFakeOKResponse(),
+			expectedListResult:   newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedGetResponse:  newFakeOKResponse(),
+			expectedGetResult:    newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					listFunc: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+						return []godo.Firewall{*test.expectedListResult}, test.expectedListResponse, test.expectedListError
+					},
+					addRulesFunc: func(ctx context.Context, fID string, rr *godo.FirewallRulesRequest) (*godo.Response, error) {
+						return test.expectedAddRulesResponse, test.expectedAddRulesError
+					},
+					updateFunc: func(ctx context.Context, fID string, fr *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedUpdateResult, test.expectedUpdateResponse, test.expectedUpdateError
+					},
+					createFunc: func(ctx context.Context, fr *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedCreateResult, test.expectedCreateResponse, test.expectedCreateError
+					},
+					getFunc: func(ctx context.Context, fID string) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedGetResult, test.expectedGetResponse, test.expectedGetError
+					},
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			err := fc.fwManager.Set(ctx, test.inboundRules)
+			if want, got := test.expectedError, err; !reflect.DeepEqual(want, got) {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
```

Comment:
We could boost the confidence in our test here by not just checking that we got the expected error (which may be none) but also asserting that the firewall being passed into the create or update call is what we passed into the `Set()` call. This can be achieved by extended our firewall test-double to store a copy of the received firewall (unless it was programmed to return an error) so that we can compare the stored result to the expected firewall.

Reply:
> the firewall being passed into the create or update call is what we passed into the Set() call

would this be in reference to the `inboundRules` getting passed into `Set()`?
```go
err := fc.fwManager.Set(ctx, test.inboundRules)
```

> This can be achieved by extended our firewall test-double to store a copy of the received firewall (unless it was programmed to return an error) so that we can compare the stored result to the expected firewall.

Comparing the rules that get passed in to `Set()`? Or is this comparing what firewall was `Set()` and then checking if it matches the expected firewall?

Also, right now I mock out the returned firewalls. So this is really just testing out that when I, for example, mocked out a create call for the godo client, that `Set()` does with that information what it should and a `Get()` should return that firewall?

Or am I misunderstanding/overcomplicating something here @timoreimann ?

Reply:
Sorry, I'll try to clarify.

The goal of `Set()` is to send a firewall object to the DigitalOcean API (for the happy path). However, right now we only test that the right error is returned, and implicitly assume that the correct firewall value was submitted if there's no error.

We can be more explicit by verifying that the expected firewall (not just inbound rules) is really passed to the _create_ or _update_ call eventually. We can even assert that the right of the two calls was invoked for a specific test case to rule out that, say, our test passes because the firewall was (successfully) submitted to _create_ when in fact in should have gone through _update_.